### ISSUE TITLE
GPU memory usage/leak varies between experiments/trials with Apex (use_amp=True) 

### ISSUE THREAD

**Issue by brucemuller** [PyTorchLightning/pytorch-lightning#616]
Using use_amp=True with hyper parameter searching results in increased GPU memory usage between experiments. So it'll go from initially 7GB usage to maybe 10GB usage between learning rate trials and then vary between that.

For example:

```
parent_parser = HyperOptArgumentParser(strategy='grid_search', add_help=False)
    
parent_parser.opt_list('--lr', default = 0.01, type=float, tunable=True, options = [0.00005, 0.00006, 0.00007, 0.00008, 0.00009, 0.0001, 0.00011, 0.00012, 0.00013, 0.00014, 0.00015])
    
    hyperparams = parent_parser.parse_args()

    for hparam_trial in hyperparams.trials(20):
        print(hparam_trial)
        main_local(hparam_trial)   
        print("NEXT TRIAL")

def main_local(hparams):

    
    model = Model(hparams)
    expname = "HPSEARCH" +  '_' + str(hparams.lr) + '_' + str(hparams.batch_size)
    tt_logger = TestTubeLogger(save_dir="/",name=expname,debug=False,create_git_tag=False)
    
    trainer = Trainer(logger = tt_logger , use_amp=True, amp_level='O2' , gpus=1,min_nb_epochs=40, max_nb_epochs=41)
      
    torch.cuda.empty_cache()
```

 - PyTorch Version: 1.3.1
 - Python version: 3.7.5
 - CUDA/cuDNN version: Driver Version: 440.33.01    CUDA Version: 10.2
 - Apex: 0.1
 - Lightning: 0.5.3.2

**Comment by jeffling** [comment]
If you put a sleep after the torch.cuda.empty_cache() (or sometime between runs) what is the GPU usage? 

Does the memory gradually increase with each run, or does it go back down after the 10gb spike?

**Comment by brucemuller** [comment]
@jeffling thanks for your reply. I did some testing with sleep after torch.cuda.empty_cache() 

With use_amp = True: GPU memory varies between 7.1gb on the initial run to 10.0gb on another.
After torch.cuda.empty_cache() it varies between 1.7gb and 5.5gb.

With use_amp = False: memory varies between 10.2gb on one run and 11gb on another. After torch.cuda.empty_cache()  the memory varies between 1.6gb and 4.3gb.

In all variations the memory is increasing between trials a few times times and then jumps down and increases gradually again. Any ideas?

**Comment by jeffling** [comment]
@brucemuller In this case, there is definitely some sort of memory leak. Have you tried directly setting the model to cpu after your run? `model.to("cpu")`

I work with a similar flow and setting that helped with our memory issues.

**Comment by brucemuller** [comment]
@jeffling thanks, I'll try that. It looks like it's not a problem with using multi-GPU/ddp but I'll keep investigating

**Comment by williamFalcon** [comment]
@jeffling @brucemuller is there anything here  we  can  fix to help?  If  so, we can  reopen this.

**Comment by brucemuller** [comment]
@williamFalcon It's a problem when not using not using distributed backend and using apex for me. 

I stopped using apex because it was killing my classification performance, which was the bigger issue.

Do you know why use_amp=true might reduce performance significantly and if there's anything we can do about it?

**Comment by williamFalcon** [comment]
apex has a few problems... soon it’ll be in pytorch and we can use that. Maybe try different optimization levels?

@ptrblck, any thoughts?

**Comment by ptrblck** [comment]
`apex.amp` wasn't designed to reinitialize models as is done e.g. for a grid search, which might yield to an increased usage of your device memory. 

@brucemuller which `opt_level` are you using how how significantly is your performance reduced?
